lib/anno_utils.pm

```diff
@@ -0,0 +1,2917 @@
+# Call pyrrolysoproteins
+# Call SEED repeat region
+# Call strep suis repeats
+# Call strep pneumo repeats
+# Call crisprs
+#
+# resolve_overlapping_features
+# Call features prophage phispy
+# retain_old_anno_for_hypotheticals
+#
+# return the workflow stages data
+#
+sub _default_genecall_workflow {
+    my $self = shift;
+    my @stages = (
+        { name => 'call_features_CDS_prodigal' },
+        { name => 'call_features_CDS_glimmer3', failure_is_not_fatal => 1,
+          glimmer3_parameters => { min_training_len => '2000'} },
+        { name => 'call_features_rRNA_SEED' },
+        { name => 'call_features_tRNA_trnascan' },
+        { name => 'call_selenoproteins', failure_is_not_fatal => 1 },
+        { name => 'call_pyrrolysoproteins', failure_is_not_fatal => 1 },
+        { name => 'call_features_repeat_region_SEED',
+              repeat_region_SEED_parameters => {} },
+        { name => 'call_features_strep_suis_repeat',
+              condition => '$genome->{scientific_name} =~ /^Streptococcus\s/' },
+        { name => 'call_features_strep_pneumo_repeat',
+              condition => '$genome->{scientific_name} =~ /^Streptococcus\s/' },
+        { name => 'call_features_crispr', failure_is_not_fatal => 1 },
+        { name => 'call_features_CDS_genemark' }
+      );
+      return { stages => \@stages };
+}
+
+sub _set_default_parameters {
+    my $self = shift;
+    return {
+        call_features_rRNA_SEED => 1,
+        call_features_tRNA_trnascan => 1,
+        call_selenoproteins => 1,
+        call_pyrrolysoproteins => 1,
+        call_features_repeat_region_SEED => 1,
+        call_features_strep_suis_repeat => 1,
+        call_features_strep_pneumo_repeat => 1,
+        call_features_crispr => 1,
+        call_features_CDS_glimmer3 => 1,
+        call_features_CDS_prodigal => 1,
+        call_features_CDS_genemark => 1,
+        annotate_proteins_kmer_v2 => 1,
+        kmer_v1_parameters => 1,
+        annotate_proteins_similarity => 1,
+        resolve_overlapping_features => 1,
+        call_features_prophage_phispy => 1,
+        retain_old_anno_for_hypotheticals => 1
+    };
+}
+
+
+sub _get_oldfunchash_oldtype_types {
+    my ($self, $in_genome) = @_;
+
+    my $oldfunchash = {};
+    my $oldtype     = {};
+    my %types = ();
+
+    for (my $i=0; $i < @{$in_genome->{features}}; $i++) {
+        my $ftr = $in_genome->{features}->[$i];
+        if (!defined($ftr->{type}) || $ftr->{type} lt '     ') {
+            if (defined($ftr->{protein_translation})) {
+                $ftr->{type} = 'gene';
+            } else {
+                $ftr->{type} = 'other';
+            }
+        }
+
+        # Reset functions in protein features to "hypothetical protein" to make them available
+        # for re-annotation in RAST service (otherwise these features will be skipped).
+        if (lc($ftr->{type}) eq "cds" || lc($ftr->{type}) eq "peg" ||
+                ($ftr->{type} eq "gene" and defined($ftr->{protein_translation}))) {
```

Comment:
it is better to stick to using either `||` and `&&` or `and` and `or`; don't mix and match because they have different levels of precedence. I would rewrite this as:
```perl
if (lc($ftr->{type}) eq "cds" || lc($ftr->{type}) eq "peg" ||
                  $ftr->{type} eq "gene" && defined $ftr->{protein_translation} ) {
```

Reply:
changed